File: Assets/FarmTap/Temple/Scripts/Controller/CharactorController.ts
import { _decorator, BoxCollider, Camera, Component, geometry, instantiate, Node, ParticleSystem, PhysicsSystem, Prefab, Quat, SkeletalAnimation, tween, Vec3, easing, find, screen, CCBoolean, Collider, RigidBody } from 'cc';
import { CharacterStateType } from '../Enum/CharacterStateType';
import { ICharacterState } from '../State/ICharacterState';
import { MoveState } from '../State/MoveState';
import { IdleState } from '../State/IdleSate';
import { ImpactState } from '../State/impactState';
import { Gamecontroller } from './Gamecontroller';
import { Sound, SoundManager } from '../Utility/SoundManager';
import { WayPoint } from '../WayPoint/WayPoint';
const { ccclass, property } = _decorator;

@ccclass('CharactorController')
export class CharactorController extends Component {
    @property(Node)
    startNode: Node = null;


    @property(Node)
    player: Node = null;
    @property
    moveSpeed: number = 10;
    @property
    safeDistance: number = 1.0;
    @property(Node)
    posEffect: Node = null;
    @property(Node)
    boom: Node = null;
    @property(Number)
    public distanceToStop: number = 0.3;
    @property(CCBoolean)
    isSheep: boolean = false;
    @property(CCBoolean)
    isCow: boolean = false;
    @property(CCBoolean)
    isPig: boolean = false;
    @property(CCBoolean)
    public isOneWay: boolean = false;

    @property(ParticleSystem)
    smokeEffect: ParticleSystem = null;
    @property(ParticleSystem)
    dustEffect: ParticleSystem = null;

    public wayPoint: WayPoint = null;

    private _currentState: ICharacterState;
    private _currentStateType: CharacterStateType;
    private _states: Map<CharacterStateType, ICharacterState>;

    public isMoving: boolean = false;
    public moveDirection: Vec3 = new Vec3();
    public startPosition: Vec3 = new Vec3();
    public distanceMoved: number = 0;
    public targetDistance: number = 0;
    private maxScanDistance: number = 30;
    public checkDistance: number = 0;

    public safePos: Vec3 = new Vec3();
    public hitPos: Vec3 = new Vec3();
    public currentPosBoom: Vec3 = new Vec3();

    private _isInImpact: boolean = false;
    public isHaveBox: boolean = true;
    public isSheepMove: boolean = false;
    public isHitWayPoint: boolean = false;
    public isChar: boolean = false;
    public isCanNotInteract: boolean = false;

    private _skeleton: SkeletalAnimation;
    public nodeClosest: Node;

    protected start(): void {
        this._skeleton = this.player.getComponent(SkeletalAnimation);
        this.initializeStates();
        this.setPivotSmoke();
        // if (this.isSheep) {
        //     this.scheduleOnce(() => {
        //         this.onClick();
        //     },1.5)
        // }
    }

    private _deltaTime: number = 0;

    public getDeltaTime(): number {
        return this._deltaTime;
    }



    update(dt: number) {
        this._deltaTime = dt;

        if (this._currentState) {
            this._currentState.update(this);
        }

        if (this._currentStateType === CharacterStateType.IDLE) return;
    }

    setPivotSmoke() {
        let angleParent = (this.node.parent.eulerAngles.y + 360) % 360;


        let angle = (this.node.eulerAngles.y + 360) % 360;
        this.dustEffect.startRotation3D = true;
        let deg = (angle + angleParent + 90) * Math.PI / 180;
        this.dustEffect.startRotationX.constant = 0;
        this.dustEffect.startRotationY.constant = 0;
        this.dustEffect.startRotationZ.constant = deg;
    }

    private initializeStates() {
        this._states = new Map();
        this._states.set(CharacterStateType.IDLE, new IdleState());
        this._states.set(CharacterStateType.RUN, new MoveState());
        this._states.set(CharacterStateType.IMPACT, new ImpactState());

        this.changeState(CharacterStateType.IDLE);
    }

    public changeState(newStateType: CharacterStateType) {
        if (this._currentStateType === newStateType) return;
        let newState = this._states.get(newStateType);
        if (!newState) return;
        if (this._currentState) {
            this._currentState.exit(this);
        }
        this._currentState = newState;
        this._currentStateType = newStateType;

        if (newStateType === CharacterStateType.IMPACT) {
            this._isInImpact = true;
        } else if (newStateType === CharacterStateType.IDLE) {
            this._isInImpact = false;
        }

        this._currentState.enter(this);
    }

    public getCurrentState(): ICharacterState {
        return this._currentState;
    }

    public getCurrentStateType(): CharacterStateType {
        return this._currentStateType;
    }

    public playAnimation(animName: string) {
        if (this._skeleton && animName) {
            this._skeleton.play(animName);
        }
        return null;
    }

    public playRandomIdleAnimation() {
        if (!this._skeleton) return;

        const idleAnimations = [
            CharacterStateType.IDLE,
            // CharacterStateType.IDLE2,
            // CharacterStateType.IDLE3
        ];

        const randomIndex = Math.floor(Math.random() * idleAnimations.length);
        const selectedAnimation = idleAnimations[randomIndex];

        this._skeleton.play(selectedAnimation);

    }
    public playIdleAnimation(useRandom: boolean = true, specificIdle?: CharacterStateType) {
        if (!this._skeleton) return;

        if (useRandom) {
            this.playRandomIdleAnimation();
        } else {
            const idleToPlay = specificIdle || CharacterStateType.IDLE;
            this._skeleton.play(idleToPlay);
        }
    }

    public handleImpactAnimationEnd() {
        if (!this._skeleton) return;
        const animationState = this._skeleton.getState(CharacterStateType.IMPACT);
        Gamecontroller.instance.playEffect(this.posEffect);
        if (animationState) {
            const duration = animationState.duration;

            this.scheduleOnce(() => {
                if (this._currentStateType === CharacterStateType.IMPACT) {
                    Gamecontroller.instance.faceEffect.playAnimationFace(this.node, "Angry");
                    this.dustEffect.play();
                    this.changeState(CharacterStateType.IDLE);
                    Gamecontroller.instance.checkCharCanMove();
                    if (!this.isSheep) return;
                    //SoundManager.Instance(SoundManager).playSound("fart");
                }
            }, duration / 8);
        }
    }

    private performSweepScan(origin: Vec3, direction: Vec3): number {
        const ray = new geometry.Ray(origin.x, origin.y, origin.z, direction.x, direction.y, direction.z);
        const mask = 0xffffffff; // Quét tất cả layer
        const queryTrigger = false;

        const hasHit = PhysicsSystem.instance.raycastClosest(ray, mask, this.maxScanDistance, queryTrigger);

        if (!hasHit) {
            Gamecontroller.instance.checkCharCanMove();
            Gamecontroller.instance.faceEffect.playAnimationFace(this.node, "Laugh");
            this.isHaveBox = false;
            this.isSheepMove = true;
            let a = this.node.getComponent(BoxCollider);
            if (a) a.destroy()
            this.scheduleOnce(() => {
                this.checkAndDestroyIfOutOfViewport();
            }, 0.1);
        }
        else {
            const result = PhysicsSystem.instance.raycastClosestResult;

            this.checkDistance = result.distance;
            let closestDistance = this.maxScanDistance;
            let foundValidHit = false;

            this.nodeClosest = result.collider.node;

            const otherController = result.collider.node.getComponent(CharactorController);
            const wayPoint = result.collider.node.parent.parent.getComponent(WayPoint);
            if (otherController) {
                if (result.distance < closestDistance) {
                    closestDistance = result.distance;
                    foundValidHit = true;
                    this.isChar = true;
                }
            } else if (wayPoint) {
                Gamecontroller.instance.checkCharCanMove();
                Gamecontroller.instance.cointEffect.playCoinEffect(this.node);
                Gamecontroller.instance.updateProgressBar();
                this.isHitWayPoint = true;
                this.wayPoint = wayPoint;
                this.node.getComponent(RigidBody).group = 16;
                //wayPoint.moveToNextWayPoint(this.node)
                if (this.isOneWay) {
                    return result.distance + 0.1;
                }
                else {
                    return result.distance + this.distanceToStop;
                }
            }
            else {
                this.node.getComponent(Collider).enabled = false;
                if (result.distance < closestDistance) {
                    closestDistance = result.distance;
                    foundValidHit = true;
                    this.scheduleOnce(() => {
                        Gamecontroller.instance.checkCharCanMove();
                        this.node.destroy();
                    }, (closestDistance / this.moveSpeed));
                }
            }

            if (foundValidHit) {
                return Math.max(0, closestDistance - this.safeDistance);
            }
        }
        return this.maxScanDistance;
    }

    public setupMovement(origin: Vec3, direction: Vec3) {
        // Reset wayPoint flag
        this.isHitWayPoint = false;

        this.moveDirection.set(direction);
        this.moveDirection.normalize();

        const safeDistance = this.performSweepScan(origin, this.moveDirection);

        const hitDistance = (safeDistance < this.maxScanDistance)
            ? safeDistance + this.safeDistance
            : this.maxScanDistance;

        this.hitPos = origin.clone().add(this.moveDirection.clone().multiplyScalar(hitDistance));
        this.safePos = origin.clone().add(this.moveDirection.clone().multiplyScalar(safeDistance));

        this.startPosition = origin.clone();
        this.distanceMoved = 0;
        this.targetDistance = hitDistance - 0.5;
    }

    public getClosestNode(nodeClose: Node) {
        if (!nodeClose || !nodeClose.getComponent(CharactorController)) return;
        if (this.isSheep) {
            SoundManager.Instance(SoundManager).playSound("Uh_1");
        }
        if (this.isCow) {
            SoundManager.Instance(SoundManager).playSound("Uh_1");
        }
        if (this.isPig) {
            SoundManager.Instance(SoundManager).playSound("pig_va_cham");
        }
        const currentPosOther = nodeClose.position.clone();
        if (this.boom != null) {
            this.currentPosBoom = this.boom.position.clone();
        }

        //Gamecontroller.instance.playAnimationFace(nodeClose, "Angry");
        if (this.node.eulerAngles.y == 0 || this.node.eulerAngles.y == 180) {

            tween(nodeClose)
                .to(0.05, { position: new Vec3(currentPosOther.x, currentPosOther.y, currentPosOther.z + 0.2) }, { easing: easing.sineOut })
                .to(0.05, { position: currentPosOther }, { easing: easing.backOut })
                .start();
            if (this.boom != null) {
                tween(this.boom)
                    .to(0.05, { position: new Vec3(this.currentPosBoom.x + 0.2, this.currentPosBoom.y, this.currentPosBoom.z) }, { easing: easing.sineOut })
                    .to(0.05, { position: this.currentPosBoom }, { easing: easing.backOut })
                    .start();
            }
        }
        else {
            tween(nodeClose)
                .to(0.05, { position: new Vec3(currentPosOther.x + 0.2, currentPosOther.y, currentPosOther.z) }, { easing: easing.sineOut })
                .to(0.05, { position: currentPosOther }, { easing: easing.backOut })
                .start();
            if (this.boom != null) {
                tween(this.boom)
                    .to(0.05, { position: new Vec3(this.currentPosBoom.x, this.currentPosBoom.y, this.currentPosBoom.z + 0.2) }, { easing: easing.sineOut })
                    .to(0.05, { position: this.currentPosBoom }, { easing: easing.backOut })
                    .start();
            }
        }

        this.nodeClosest = nodeClose;
    }
    onClick() {
        if (this._currentStateType !== CharacterStateType.IDLE) {
            return;
        }

        const direction = new Vec3();
        Vec3.negate(direction, this.startNode.forward);
        const origin = this.startNode.worldPosition.clone();

        // Kiểm tra xem có thể di chuyển không
        const canMove = this.checkMove();
        if (!canMove) {
            console.log("Movement blocked by obstacle with group 16");
            return;
        }

        this.setupMovement(origin, direction);
        this.changeState(CharacterStateType.RUN);

    }

    public stopMovement() {
        this.isMoving = false;
    }

    public checkMove() {
        const origin = this.startNode.worldPosition.clone();
        const direction = new Vec3();
        Vec3.negate(direction, this.startNode.forward);
        direction.normalize();

        const ray = new geometry.Ray(origin.x, origin.y, origin.z, direction.x, direction.y, direction.z);
        const mask = 0xffffffff;
        const queryTrigger = false;

        const hasHit = PhysicsSystem.instance.raycast(ray, mask, this.maxScanDistance, queryTrigger);

        if (hasHit) {
            const results = PhysicsSystem.instance.raycastResults;
            for (let i = 0; i < results.length; i++) {
                const result = results[i];

                if (result.collider.node.getComponent(RigidBody)) {
                    const rigidBody = result.collider.node.getComponent(RigidBody);

                    if (rigidBody.group === 16) {
                        return false;
                    }
                }
            }

            return true;
        } else {
            return true;
        }
    }

    public get isInImpact(): boolean {
        return this._isInImpact;
    }

    private isNodeOutOfViewport(): boolean {
        const cameraNode = find("Main Camera");
        if (!cameraNode) return false;

        const camera = cameraNode.getComponent(Camera);
        if (!camera) return false;
        const worldPos = this.node.worldPosition;
        const screenPos = camera.worldToScreen(worldPos);
        const { width, height } = screen.windowSize;
        const margin = -30;
        return screenPos.x < -margin ||
            screenPos.x > width + margin ||
            screenPos.y < -margin ||
            screenPos.y > height + margin;
    }

    private checkAndDestroyIfOutOfViewport(): void {
        if (this.isNodeOutOfViewport()) {
            //console.log("Node is out of viewport, destroying...");
            Gamecontroller.instance.checkCharCanMove();
            this.node.destroy();
        } else {
            this.scheduleOnce(() => {
                this.checkAndDestroyIfOutOfViewport();
            });
        }
    }

    public canMove(): boolean {
        if (!this.node || !this.node.isValid) {
            return false;
        }
        const origin = this.startNode.worldPosition.clone();
        const direction = new Vec3();
        Vec3.negate(direction, this.startNode.forward);
        direction.normalize();

        const ray = new geometry.Ray(origin.x, origin.y, origin.z, direction.x, direction.y, direction.z);
        const mask = 0xffffffff & ~(1 << 4);
        const queryTrigger = false;

        const hasHit = PhysicsSystem.instance.raycastClosest(ray, mask, this.maxScanDistance, queryTrigger);

        if (!hasHit) {
            return true;
        }

        const result = PhysicsSystem.instance.raycastClosestResult;
        const currentDistance = result.distance;
        return currentDistance >= this.safeDistance;
    }


}

File: Assets/FarmTap/Scripts/Player/CharactorController.cs
using DG.Tweening;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class CharactorController : MonoBehaviour
{
    [Header("References")]
    [SerializeField] private Transform startNode;
    [SerializeField] private GameObject player;
    [SerializeField] public GameObject posEffect;
    [SerializeField] public GameObject trails;
    [SerializeField] private ParticleSystem smokeEffect;
    [SerializeField] public ParticleSystem dustEffect;

    [Header("Movement Settings")]
    [SerializeField] private float moveSpeed = 10f;
    [SerializeField] private float safeDistance = 1.0f;

    [Header("Character Type")]
    [SerializeField] private bool isSheep = false;
    [SerializeField] private bool isCow = false;
    [SerializeField] private bool isPig = false;
    [SerializeField] public bool isOneWay = false;

    // Public properties
    public ParticleSystem DustEffect => dustEffect;

    // State management
    private ICharacterState _currentState;
    private CharacterStateType _currentStateType;
    private Dictionary<CharacterStateType, ICharacterState> _states;

    // Movement variables
    public bool IsMoving { get; set; }
    public Vector3 MoveDirection { get; private set; }
    public Vector3 StartPosition { get; private set; }
    public float DistanceMoved { get; set; }
    public float TargetDistance { get; private set; }
    public float CheckDistance { get; private set; }

    // Position tracking
    public Vector3 SafePos { get; private set; }
    public Vector3 HitPos { get; private set; }
    public Vector3 CurrentPosBoom { get; private set; }

    // Flags
    private bool _isInImpact = false;
    public bool IsHaveBox { get; set; } = true;
    public bool IsSheepMove { get; set; } = false;
    public bool IsHitWayPoint { get; set; } = false;
    public bool IsHitGateCheck { get; set; } = false;
    public bool IsChar { get; set; } = false;
    public bool IsCanNotInteract { get; set; } = false;

    // Private variables
    private float maxScanDistance = 30f;
    private Animator _animator;
    public GameObject NodeClosest { get; private set; }
    private float _deltaTime;

    void Start()
    {
        _animator = player.GetComponent<Animator>();

        InitializeStates();
        SetPivotSmoke();
    }

    void Update()
    {
        _deltaTime = Time.deltaTime;

        if (_currentState != null)
        {
            _currentState.Update(this);
        }

        if (_currentStateType == CharacterStateType.Idle) return;
    }

    public float GetDeltaTime()
    {
        return _deltaTime;
    }
    public float GetMoveSpeed()
    {
        return moveSpeed;
    }

    private void SetPivotSmoke()
    {
        if (dustEffect == null) return;

        float angleParent = (transform.parent.eulerAngles.y + 360) % 360;
        float angle = (transform.eulerAngles.y + 360) % 360;

        float deg = (angle + angleParent - 90) * Mathf.Deg2Rad;
        var main = dustEffect.main;
        main.startRotation3D = true;
        main.startRotationX = 0;
        main.startRotationY = 0;
        main.startRotationZ = deg;
    }

    private void InitializeStates()
    {
        _states = new Dictionary<CharacterStateType, ICharacterState>();
        _states.Add(CharacterStateType.Idle, new IdleState());
        _states.Add(CharacterStateType.Run, new MoveState());
        _states.Add(CharacterStateType.Impact, new ImpactState());

        ChangeState(CharacterStateType.Idle);
    }

    public void ChangeState(CharacterStateType newStateType)
    {
        // Exit current state
        if (_currentState != null)
        {
            _currentState.Exit(this);
        }

        // Change to new state
        _currentStateType = newStateType;
        _currentState = _states[newStateType];

        // Update animator parameters
        UpdateAnimatorParameters(newStateType);

        // Enter new state
        _currentState.Enter(this);
    }

    private void UpdateAnimatorParameters(CharacterStateType stateType)
    {
        if (_animator == null) return;

        // Reset all animation parameters to false
        _animator.SetBool("Idle", false);
        _animator.SetBool("Run", false);
        _animator.SetBool("Impact", false);

        // Set the current state parameter to true
        switch (stateType)
        {
            case CharacterStateType.Idle:
                _animator.SetBool("Idle", true);
                _isInImpact = false;
                break;
            case CharacterStateType.Run:
                _animator.SetBool("Run", true);
                _isInImpact = false;
                break;
            case CharacterStateType.Impact:
                _animator.SetBool("Impact", true);
                _isInImpact = true;
                break;
        }
    }

    public CharacterStateType GetCurrentStateType()
    {
        return _currentStateType;
    }

    private float PerformSweepScan(Vector3 origin, Vector3 direction)
    {
        RaycastHit hit;
        int mask = unchecked((int)0xffffffff);

        bool hasHit = Physics.Raycast(origin, direction, out hit, maxScanDistance, mask);
        Debug.DrawRay(origin, direction * maxScanDistance, Color.red, 2.0f);
        if (!hasHit)
        {
            Debug.Log("No hit detected in PerformSweepScan.");
            IsHaveBox = false;
            IsSheepMove = true;

            BoxCollider boxCollider = GetComponent<BoxCollider>();
            if (boxCollider != null) Destroy(boxCollider);

            StartCoroutine(CheckAndDestroyIfOutOfViewportCoroutine(0.1f));
            return maxScanDistance;
        }
        else
        {
            Debug.Log("Hit detected in PerformSweepScan.");
            CheckDistance = hit.distance;
            float closestDistance = maxScanDistance;
            bool foundValidHit = false;

            NodeClosest = hit.collider.gameObject;

            CharactorController otherController = hit.collider.GetComponent<CharactorController>();
            GateCheck gateCheck = hit.collider.GetComponent<GateCheck>();
            if (otherController != null)
            {
                if (hit.distance < closestDistance)
                {
                    closestDistance = hit.distance;
                    foundValidHit = true;
                    IsChar = true;
                }
            }
            else if (gateCheck) 
            {
                Debug.Log("gateCheck");
                closestDistance = hit.distance;
                return closestDistance;
            }
            else
            {
                Collider collider = GetComponent<Collider>();
                if (collider != null) collider.enabled = false;

                if (hit.distance < closestDistance)
                {
                    closestDistance = hit.distance;
                    foundValidHit = true;

                    float destroyDelay = closestDistance / moveSpeed;
                    StartCoroutine(DestroyAfterDelay(destroyDelay));
                }
            }

            if (foundValidHit)
            {
                return Mathf.Max(0, closestDistance - safeDistance);
            }
        }

        return maxScanDistance;
    }

    private IEnumerator DestroyAfterDelay(float delay)
    {
        yield return new WaitForSeconds(delay);
        Destroy(gameObject);
    }

    public void SetupMovement(Vector3 origin, Vector3 direction)
    {
        IsHitWayPoint = false;
        MoveDirection = direction.normalized;

        float safeDistanceResult = PerformSweepScan(origin, MoveDirection);

        float hitDistance = (safeDistanceResult < maxScanDistance)
            ? safeDistanceResult + safeDistance
            : maxScanDistance;

        HitPos = origin + MoveDirection * hitDistance;
        SafePos = origin + MoveDirection * safeDistanceResult;

        StartPosition = origin;
        DistanceMoved = 0;
        TargetDistance = hitDistance - 0.5f;
    }

    public void GetClosestNode(GameObject nodeClose)
    {
        if (nodeClose == null || nodeClose.GetComponent<CharactorController>() == null) return;

        // Play sound based on character type
        if (isSheep)
        {
            // SoundManager.Instance.PlaySound("Sheepbaa");
        }
        else if (isCow)
        {
            // SoundManager.Instance.PlaySound("Uh_1");
        }
        else if (isPig)
        {
            // SoundManager.Instance.PlaySound("pig_va_cham");
        }

        Vector3 currentPosOther = nodeClose.transform.position;

        // Apply impact animation using DOTween or coroutines
       ApplyImpactAnimation(nodeClose);

        NodeClosest = nodeClose;
    }

    private void ApplyImpactAnimation(GameObject nodeClose)
    {
        Vector3 currentPosOther = nodeClose.transform.position;
        Vector3 offset = Vector3.zero;
        if(Mathf.Approximately(transform.eulerAngles.y , 0) || Mathf.Approximately(transform.eulerAngles.y, 180))
        {
            offset = new Vector3(0, 0, 0.2f);
        }
        else
        {
            offset = new Vector3(0.2f, 0, 0);   
        }

        float duration = 0.1f;
        nodeClose.transform.DOKill();
        Sequence seq = DOTween.Sequence();
        seq.Append(nodeClose.transform.DOMove(currentPosOther + offset, duration).SetEase(Ease.OutQuad));
        seq.Append(nodeClose.transform.DOMove(currentPosOther, duration).SetEase(Ease.InQuad));
    }

    public void OnClick()
    {
        if (_currentStateType != CharacterStateType.Idle)
        {
            return;
        }

        Vector3 direction = startNode.forward;
        Vector3 origin = startNode.position;

        SetupMovement(origin, direction);
        ChangeState(CharacterStateType.Run);
    }

    public void StopMovement()
    {
        IsMoving = false;
    }

    public bool CheckMove()
    {
        Vector3 origin = startNode.position;
        Vector3 direction = -startNode.forward;
        direction.Normalize();

        RaycastHit[] hits = Physics.RaycastAll(origin, direction, maxScanDistance);

        if (hits.Length > 0)
        {
            foreach (RaycastHit hit in hits)
            {
                Rigidbody rb = hit.collider.GetComponent<Rigidbody>();
                if (rb != null)
                {
                    // Check if rigidbody is in layer 16 (equivalent to group 16)
                    if (hit.collider.gameObject.layer == 4)
                    {
                        float distance = hit.distance;
                        if (distance < 4)
                        {
                            return false;
                        }
                    }
                }
            }
            return true;
        }
        else
        {
            return true;
        }
    }

    public bool IsInImpact
    {
        get { return _isInImpact; }
    }

    private bool IsNodeOutOfViewport()
    {
        Camera mainCamera = Camera.main;
        if (mainCamera == null) return false;

        Vector3 viewportPos = mainCamera.WorldToViewportPoint(transform.position);
        float margin = -30f / Screen.width;

        return viewportPos.x < -margin ||
               viewportPos.x > 1 + margin ||
               viewportPos.y < -margin ||
               viewportPos.y > 1 + margin;
    }

    private IEnumerator CheckAndDestroyIfOutOfViewportCoroutine(float delay)
    {
        yield return new WaitForSeconds(delay);

        if (IsNodeOutOfViewport())
        {
            Destroy(gameObject);
        }
        else
        {
            StartCoroutine(CheckAndDestroyIfOutOfViewportCoroutine(0.1f));
        }
    }

    public bool CanMove()
    {
        if (gameObject == null) return false;

        Vector3 origin = startNode.position;
        Vector3 direction = -startNode.forward;
        direction.Normalize();

        RaycastHit hit;
        int mask = ~(1 << 4); // All layers except layer 4

        bool hasHit = Physics.Raycast(origin, direction, out hit, maxScanDistance, mask);

        if (!hasHit)
        {
            return true;
        }

        float currentDistance = hit.distance;
        return currentDistance >= safeDistance + 1;
    }
}